software/libdvi/tmds_encode.c

```diff
@@ -162,3 +162,70 @@ void __not_in_flash_func(tmds_encode_data_channel_fullres_16bpp)(const uint32_t
 	interp_restore(interp1_hw, &interp1_save);
 #endif
 }
+
+// This takes a 16-bit (RGB 565) colour palette and makes palettes of TMDS symbols suitable
+// for performing fullres encode.
+// The TMDS palette buffer should be 6 * n_palette words long.
+// n_palette must be a power of 2 <= 256.
+void tmds_setup_palette_symbols(const uint16_t *palette, uint32_t *tmds_palette, size_t n_palette) {
+	uint32_t* tmds_palette_blue = tmds_palette;
+	uint32_t* tmds_palette_green = tmds_palette + 2 * n_palette;
+	uint32_t* tmds_palette_red = tmds_palette + 4 * n_palette;
+	for (int i = 0; i < n_palette; ++i) {
+		uint16_t blue = (palette[i] << 1) & 0x3e;
+		uint16_t green = (palette[i] >> 5) & 0x3f;
+		uint16_t red = (palette[i] >> 10) & 0x3e;
+		tmds_palette_blue[i] = tmds_table_fullres_x[blue];
+		tmds_palette_blue[i + n_palette] = tmds_table_fullres_x[64 + blue];
+		tmds_palette_green[i] = tmds_table_fullres_x[green];
+		tmds_palette_green[i + n_palette] = tmds_table_fullres_x[64 + green];
+		tmds_palette_red[i] = tmds_table_fullres_x[red];
+		tmds_palette_red[i + n_palette] = tmds_table_fullres_x[64 + red];
+	}
+}
+
+// Encode palette data for all 3 channels.
+// pixbuf is an array of n_pix 8-bit wide pixels containing palette values (32-bit word aligned)
+// tmds_palette is a palette of TMDS symbols produced by tmds_setup_palette_symbols
+// symbuf is 3*n_pix 32-bit words, this function writes the symbol values for each of the channels to it.
+void __not_in_flash_func(tmds_encode_palette_data)(const uint32_t *pixbuf, const uint32_t *tmds_palette, uint32_t *symbuf, size_t n_pix, uint32_t palette_bits) {
+	uint core = get_core_num();
+#if !TMDS_FULLRES_NO_INTERP_SAVE
+	interp_hw_save_t interp0_save, interp1_save;
+	interp_save(interp0_hw, &interp0_save);
+	interp_save(interp1_hw, &interp1_save);
+#endif
+
+	interp0_hw->base[2] = (uint32_t)tmds_palette;
+	interp1_hw->base[2] = (uint32_t)tmds_palette;
+	interp0_hw->ctrl[0] = 0x40u | ((palette_bits + 1) << SIO_INTERP0_CTRL_LANE0_MASK_MSB_LSB);
```

Comment:
Could you clean up the magic numbers please (in return I will clean up the ones I lazily left in vista/main.c)

Reply:
Ah yes, these evolved from the ones in vista, I'll fix them up.